### ISSUE TITLE
Security Vulnerability in Undertow Core dependency CVE-2018-1048

### ISSUE THREAD

**Issue by RothAndrew** [jhipster/generator-jhipster#9091]
##### **Overview of the issue**

Static Security Analysis tool (FOSSA.io) is reporting vulnerability CVE-2018-1048 in the Undertow Core dependency, which is a sub-depencency of `org.springframework.boot:spring-boot-starter-undertow`. Unfortunately it doesn't offer a resolution, but I wanted to get it on your radar. The version of Undertow Core being used in the JHipster app I generated is `1.4.26.Final`.

Description:

> It was found that the AJP connector in undertow, as shipped in Jboss EAP 7.1.0.GA, does not use the ALLOW_ENCODED_SLASH option and thus allow the the slash / anti-slash characters encoded in the url which may lead to path traversal and result in the information disclosure of arbitrary local files.

References:

* https://bugzilla.redhat.com/show_bug.cgi?id=1534343
* https://nvd.nist.gov/vuln/detail/CVE-2018-1048

##### **Motivation for or Use Case**

n/a

##### **Reproduce the error**

Generate a JHipster app. I generated a Microservice Gateway, but I'm guessing any JHipster app will be affected.

##### **Related issues**

Couldn't find any calling out this CVE

##### **Suggest a Fix**

Not sure yet, usually FOSSA just gives a new version of the dependency to upgrade to, but it isn't saying there is one this time.

##### **JHipster Version(s)**

5.7.2

##### **JHipster configuration**

```
$ jhipster info
INFO! Using JHipster version installed locally in current project's node_modules
INFO! Executing jhipster:info
INFO! Options: from-cli: true
Welcome to the JHipster Information Sub-Generator

##### **JHipster Version(s)**


helloworldgateway@0.0.0 /home/redacted/hello-world-gateway
└── generator-jhipster@5.7.2




##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**


<details>
<summary>.yo-rc.json file</summary>
<pre>
{
    "generator-jhipster": {
        "promptValues": {
            "packageName": "com.redacted.helloworld.gateway",
            "nativeLanguage": "en"
        },
        "jhipsterVersion": "5.7.2",
        "applicationType": "gateway",
        "baseName": "helloworldgateway",
        "packageName": "com.redacted.helloworld.gateway",
        "packageFolder": "com/redacted/helloworld/gateway",
        "serverPort": "8080",
        "authenticationType": "jwt",
        "cacheProvider": "hazelcast",
        "enableHibernateCache": true,
        "websocket": false,
        "databaseType": "sql",
        "devDatabaseType": "h2Disk",
        "prodDatabaseType": "mysql",
        "searchEngine": false,
        "messageBroker": false,
        "serviceDiscoveryType": false,
        "buildTool": "gradle",
        "enableSwaggerCodegen": false,
        "jwtSecretKey": "redacted",
        "clientFramework": "angularX",
        "useSass": true,
        "clientPackageManager": "npm",
        "testFrameworks": ["cucumber"],
        "jhiPrefix": "jhi",
        "otherModules": [],
        "enableTranslation": true,
        "nativeLanguage": "en",
        "languages": ["en"]
    }
}

</pre>
</details>


##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

<details>
<summary>JDL entity definitions</summary>

<pre>

</pre>
</details>


##### **Environment and Tools**

openjdk version "1.8.0_191"
OpenJDK Runtime Environment (build 1.8.0_191-8u191-b12-0ubuntu0.18.04.1-b12)
OpenJDK 64-Bit Server VM (build 25.191-b12, mixed mode)

git version 2.17.1

node: v10.15.0

npm: 6.4.1

yarn: 1.13.0
```

##### **Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

I haven't generated any entities yet.

##### **Browsers and Operating System**

```
$ cat /etc/os-release
NAME="Linux Mint"
VERSION="19 (Tara)"
ID=linuxmint
ID_LIKE=ubuntu
PRETTY_NAME="Linux Mint 19"
VERSION_ID="19"
HOME_URL="https://www.linuxmint.com/"
SUPPORT_URL="https://forums.ubuntu.com/"
BUG_REPORT_URL="http://linuxmint-troubleshooting-guide.readthedocs.io/en/latest/"
PRIVACY_POLICY_URL="https://www.linuxmint.com/"
VERSION_CODENAME=tara
UBUNTU_CODENAME=bionic
```

-   [x] Checking this box is mandatory (this is just to show you read everything)

**Comment by RothAndrew** [comment]
More references, for posterity :)

* https://issues.jboss.org/browse/UNDERTOW-1295
* https://issues.jboss.org/browse/UNDERTOW-1245

**Comment by jdubois** [comment]
Otherwise we can migrate to Tomcat, that's no big deal (but it's a bit slower).

**Comment by krusche** [comment]
Is there a reason, why jhipster does not yet use undertow 2.x?

**Comment by gmarziou** [comment]
I just looked at the code of the  `decode` method in `AjpRequestParser` as mentioned in https://issues.jboss.org/browse/UNDERTOW-1245

It does use ALLOW_ENCODED_SLASH option which is false by default at least in version 1.4.26.Final used by JHipster..

~~~java
private String decode(String url, final boolean containsUrlCharacters) throws UnsupportedEncodingException {
        if (doDecode && containsUrlCharacters) {
            try {
                if(decodeBuffer == null) {
                    decodeBuffer = new StringBuilder();
                }
                return URLUtils.decode(url, this.encoding, allowEncodedSlash, false, decodeBuffer);
~~~

Also if you look at the way Spring Boot configures Undertow in org.springframework.boot.web.embedded.undertow.UndertowServletWebServerFactory#createBuilder() you'll see that it calls only `addHttpListener()` and not `addAjpListener()`.

So for me, we are not vulnerable because Spring Boot does not configure Ajp in Undertow and even if it did Undertow has a safe default value.

**Comment by gmarziou** [comment]
@krusche simply because Undertow 2.x is included in Spring Boot 2.1 and as of now JHipster 5.7.2 uses Spring Boot 2.0.x

**Comment by RothAndrew** [comment]
oops wrong button. However, I'm fine with this being closed now after Gael's investigation.

**Comment by jdubois** [comment]
Yes this should be closed:

- We don't use AJP by default, and I should have mentioned it from the start, I've been telling customers to avoid AJP for at least 10 years...
- This will be fixed with Spring Boot 2.1, so that will be fixed automatically in our next major release
